### PR TITLE
[RPC] Add utility signinput

### PR DESCRIPTION
**Pull request by NicolasDorier**
An alternative to https://github.com/bitcoin/bitcoin/pull/11653 , an utility to sign an input. This is very useful when dealing with several bitcoin-based crypto currencies (typically a cross chain swap tool), as several crypto currencies signs in different ways.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
I agree it could be useful and I once did a similar thing for debugging. It's pretty expertish, though very useful.
This is pure utility (right?) so the question comes up why not to add this into `bitcoin-tx`.

**Comment by NicolasDorier** [comment]
@jonasschnelli, maybe I misunderstand the purpose of `bitcoin-tx`, but I need it to be a RPC method so I can easily make a cross chain swap tool across bitcoin like currencies without knowing the specific way of signing or how to serialize a transaction on each chain. 

My tool has already RPC access to each chain configured. Using this utility via `bitcoin-tx` programmatically make things harder, as I then need to ask users to not only configure RPC settings in my tools, but also the path to `bitcoin-tx` utility.
`bitcoin-tx` is adapted for making transactions manually. This is not what I am looking for.

**Comment by dcousens** [comment]
Is this simply `signrawtransaction`,  but only returning the signature(s) instead of the transaction hex?

**Comment by NicolasDorier** [comment]
`signrawtransaction` works only if you sign well known scripts as it use `ProduceSignature`. So in my case, it does not returns me the signature, even in the tx hex it returns me.

File: src/rpc/server.cpp [hunk 1]
```diff
@@ -131,6 +131,30 @@ uint256 ParseHashV(const UniValue& v, std::string strName)
...
+    static std::map<std::string, int> mapSigHashValues = {
+        { std::string("ALL"), int(SIGHASH_ALL) },
+        { std::string("ALL|ANYONECANPAY"), int(SIGHASH_ALL | SIGHASH_ANYONECANPAY) },
+        { std::string("NONE"), int(SIGHASH_NONE) },
+        { std::string("NONE|ANYONECANPAY"), int(SIGHASH_NONE | SIGHASH_ANYONECANPAY) },
+        { std::string("SINGLE"), int(SIGHASH_SINGLE) },
+        { std::string("SINGLE|ANYONECANPAY"), int(SIGHASH_SINGLE | SIGHASH_ANYONECANPAY) },
+    };
+    auto it = mapSigHashValues.find(sigHash);
+    if (it != mapSigHashValues.end())
```
Comment (promag) [inline src/rpc/server.cpp]: Early return (throw):
```cpp
if (it == end) throw ...;

return it->second;
```
This way `nHashType` can be removed.

File: src/rpc/server.cpp [hunk 2]
```diff
@@ -131,6 +131,30 @@ uint256 ParseHashV(const UniValue& v, std::string strName)
     result.SetHex(strHex);
     return result;
 }
+
+int ParseSigHash(const std::string& sigHash, const std::string& parameterName)
+{
+    int nHashType = 0;
+    static std::map<std::string, int> mapSigHashValues = {
```
Comment (promag) [inline src/rpc/server.cpp]: Const and snake case:
```cpp
static const std::map<std::string, int> sighash_values = ...
```

File: src/rpc/server.cpp [hunk 3]
```diff
@@ -131,6 +131,30 @@ uint256 ParseHashV(const UniValue& v, std::string strName)
     result.SetHex(strHex);
     return result;
 }
+
+int ParseSigHash(const std::string& sigHash, const std::string& parameterName)
```
Comment (promag) [inline src/rpc/server.cpp]: Nit, snake case:
```cpp
int ParseSigHash(const std::string& sighash, const std::string& parameter_name)

File: src/rpc/misc.cpp [hunk 4]
```diff
@@ -320,6 +320,89 @@ UniValue createmultisig(const JSONRPCRequest& request)
...
+        );
+
+    RPCTypeCheck(request.params,
+    {
+        UniValue::VSTR, // tx
+        UniValue::VSTR, // scriptCode
+        UniValue::VNUM, // amount
+        UniValue::VSTR, // sigVersion
+        UniValue::VNUM, // inputIndex
+        UniValue::VSTR // privatekey
```
Comment (promag) [inline src/rpc/misc.cpp]: Nit, align comments?

**Review by promag — COMMENTED** [review]
The code could be updated according to dev notes.

**Comment by NicolasDorier** [comment]
nits addressed

**Comment by NicolasDorier** [comment]
will reopen the day I need it again. I workedaround this for now.

**Pull request closed without merging**